libsolidity/ast/Types.cpp

```diff
@@ -1652,22 +1652,40 @@ bool ArrayType::validForCalldata() const
 
 bigint ArrayType::unlimitedCalldataEncodedSize(bool _padded) const
 {
-	if (isDynamicallySized())
-		return 32;
-	// Array elements are always padded.
-	bigint size = bigint(length()) * (isByteArray() ? 1 : baseType()->calldataEncodedSize(true));
+	bigint size = bigint(length()) * calldataStride();
 	if (_padded)
 		size = ((size + 31) / 32) * 32;
 	return size;
 }
 
 unsigned ArrayType::calldataEncodedSize(bool _padded) const
 {
+	solAssert(!isDynamicallyEncoded(), "");
 	bigint size = unlimitedCalldataEncodedSize(_padded);
 	solAssert(size <= numeric_limits<unsigned>::max(), "Array size does not fit unsigned.");
 	return unsigned(size);
 }
 
+unsigned ArrayType::calldataHeadIncrement() const
+{
+	if (isDynamicallyEncoded())
+		return 32;
+	else
+		return calldataEncodedSize(true);
+}
+
+unsigned ArrayType::calldataEncodedTailSize() const
+{
+	solAssert(isDynamicallyEncoded(), "");
+	if (isDynamicallySized())
+		// We do not know the dynamic length itself, but at least the uint256 containing the
+		// length must still be present.
+		return 32;
+	bigint size = unlimitedCalldataEncodedSize(false);
```

Comment:
It feels strange to explicitly tell `unlimitedCalldataEncodedSize` to avoid padding. Not sure if you want to emphasize something? If I understand right it actually doesn't make any difference because tail pointers are exactly 32 bytes.

Reply:
It doesn't make a difference, yes - the only case in which this would not by the same is ``string`` and ``bytes``, but those are dynamically sized, so for those this function will return ``32`` anyways.
I'd still consider unpadded more correct - this function returns the minimal tail size used for bounds checks during calldata validation and I'd not consider the encoding of an array with missing padding in the end an invalid encoding... but yeah - such a case doesn't exist anyways.